### PR TITLE
Study #ano now works

### PR DESCRIPTION
**Pull request by jtexeira**

### CODE REVIEW

File: src/channels.rs [hunk 1]
```diff
@@ -24,12 +25,21 @@ impl MiEI {
...
 
-    pub fn get_role_id(&self, role_name: &str) -> Option<RoleId> {
-        self.courses
-            .values()
-            .filter_map(|x| x.courses.get(role_name))
+    pub fn get_role_id(&self, role_name: &str) -> Vec<RoleId> {
+        let years = &self.courses;
+        let year = Regex::new("ano")
+            .unwrap()
+            .split(role_name)
```
Comment (Mendess2526) [inline src/channels.rs]: The standard library can already handle this: https://doc.rust-lang.org/std/string/struct.String.html#method.split
no need to add the regex crate.

You could instead use the regex to determine the type of role request:
```
if /* regex for year */ {
    years
} else {
    course
}
```

File: src/channels.rs [hunk 2]
```diff
@@ -24,12 +25,21 @@ impl MiEI {
...
             .next()
+            .unwrap();
+        let course = self.courses.get(year);
+        match course {
+            Some(x) => x.courses.values().collect::<Vec<&Course>>(),
+            None => years.values().filter_map(|x| x.courses.get(role_name)).collect::<Vec<&Course>>(),
+        }
+            .iter()
```
Comment (Mendess2526) [inline src/channels.rs]: Don't allocate the vector twice. Just return the iterator form.

File: src/channels.rs [hunk 3]
```diff
@@ -24,20 +25,33 @@ impl MiEI {
...
-        self.courses
-            .values()
-            .filter_map(|x| x.courses.get(role_name))
-            .next()
-            .map(|x| x.role.parse::<RoleId>().unwrap())
+    pub fn get_role_id(&self, role_name: &str) -> Vec<RoleId> {
+        let years = &self.courses;
+        let regex = Regex::new("([0-9]+)ANO([0-9]+)(SEMESTRE|SEM)").unwrap();
+        let year_regex = Regex::new("([0-9])+ANO").unwrap();
+        if regex.is_match(role_name) {
```
Comment (Mendess2526) [inline src/channels.rs]: `Regex::captures` returns an option. If you use `if let` you don't need to parse `role_name` twice and you don't need to `unwrap`
```
if let Some(splits) = regex.captures(role_name) {
    // use `splits`
}
```

File: src/commands.rs [hunk 4]
```diff
@@ -21,7 +21,8 @@ pub fn study(ctx: &mut Context, msg: &Message, mut args: Args) -> CommandResult
     let ids = args
         .iter::<String>()
         .filter_map(Result::ok)
-        .filter_map(|x| roles.get_role_id(&x))
+        .map(|x| roles.get_role_id(&x.to_uppercase()))
```
Comment (Mendess2526) [inline src/commands.rs]: Regex can be case insensitive, using this allocates a new string for every role.
https://docs.rs/regex/1.2.1/regex/#grouping-and-flags

File: src/channels.rs [hunk 5]
```diff
@@ -24,20 +25,33 @@ impl MiEI {
...
-    pub fn get_role_id(&self, role_name: &str) -> Option<RoleId> {
-        self.courses
-            .values()
-            .filter_map(|x| x.courses.get(role_name))
-            .next()
-            .map(|x| x.role.parse::<RoleId>().unwrap())
+    pub fn get_role_id(&self, role_name: &str) -> Vec<RoleId> {
+        let years = &self.courses;
+        let regex = Regex::new("([0-9]+)ANO([0-9]+)(SEMESTRE|SEM)").unwrap();
+        let year_regex = Regex::new("([0-9])+ANO").unwrap();
```
Comment (Mendess2526) [inline src/channels.rs]: Please use a `lazy_static`: https://docs.rs/regex/1.2.1/regex/#example-avoid-compiling-the-same-regex-in-a-loop

File: src/commands.rs (hunk 4 above)
Comment (jtexeira) [reply src/commands.rs]: This is for the individual courses, for instance, ``$study ea``. In this case, this string is directly used to search in the hashmap, where the keys all in uppercase. I guess I could use this just when this is the case

**Pull request merged**